### PR TITLE
Add compiler error when compiling with an outdated cuDNN

### PR DESCRIPTION
**Pull request by zou3519**
Fixes https://github.com/pytorch/pytorch/issues/3093

### Test Plan
`python setup.py build install` (I have cudnn 6 installed). Asserted that no error was thrown.

Changed the macro to `if CUDNN_MAJOR < 7` and recompiled with `python setup.py build install`. Asserted that a compiler error occurred: 

image

(It's okay that it says "CuDNN v6 found" because the macro was changed to < 7 for testing purposes)

### CODE REVIEW

File: torch/csrc/cudnn/cudnn-wrapper.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#ifndef THP_CUDNN_CUDNN_WRAPPER_INC
+#define THP_CUDNN_CUDNN_WRAPPER_INC
+
+#include <cudnn.h>
+#if CUDNN_MAJOR < 6
+#error CuDNN v5 and lower not supported. Please update or build without it.
```
Comment (soumith) [inline torch/csrc/cudnn/cudnn-wrapper.h]: "Please update or build without it" -> "Please install CuDNN v6 or greater. You can get the latest version of CuDNN from https://developer.nvidia.com/cudnn"

File: torch/csrc/cudnn/cudnn-wrapper.h [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#ifndef THP_CUDNN_CUDNN_WRAPPER_INC
```
Comment (colesbury) [inline torch/csrc/cudnn/cudnn-wrapper.h]: Just use `#pragma once` in new header files.

**Review by colesbury — APPROVED** [review]
lgtm

File: torch/csrc/cudnn/cudnn-wrapper.h (hunk 1 above)
Comment (soumith) [reply torch/csrc/cudnn/cudnn-wrapper.h]: actually, better:
"CuDNN v5 and lower not supported...." -> "CuDNN v" #CUDNN_MAJOR " found, but need atleast CuDNN v6. Get the latest version of CuDNN from https://developer.nvidia.com/cudnn"

Comment (zou3519) [reply torch/csrc/cudnn/cudnn-wrapper.h]: #error doesn't support using macros after it (it just dumps whatever is after it literally) but after some preprocessor hackery I think I have something that works like this...

File: torch/csrc/cudnn/cudnn-wrapper.h [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include <cudnn.h>
+
+#define STRINGIFY(x) #x
+#define STRING(x) STRINGIFY(x)
+
+#if CUDNN_MAJOR < 6
+#pragma message ("CuDNN v" STRING(CUDNN_MAJOR) " found, but need at least CUDNN v6. You can get the latest version of CUDNN from https://developer.nvidia.com/cudnn")
+#error "Old version of CuDNN found but need at least CuDNN v6. You can get the latest version of CuDNN from https://developer.nvidia.com/cudnn"
```
Comment (soumith) [inline torch/csrc/cudnn/cudnn-wrapper.h]: it looks like a botched commit?
#pragma message and #error contain a lot of duplicate wording.

**Comment by zou3519** [comment]
Closing in favor of performing the checks purely in python in #3114 because setup.py is always run to start the compilation.

**Pull request closed without merging**